core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/domain/solution/cloner/DeepCloningFieldCloner.java

```diff
@@ -1,37 +1,81 @@
-    public <C> void clone(DeepCloningUtils deepCloningUtils, Field field, Class<? extends C> instanceClass, C original, C clone,
-            Consumer<Object> deferredValueConsumer) {
-        Object originalValue = FieldCloner.getFieldValue(original, field);
-        if (isDeepCloneField(deepCloningUtils, field, instanceClass, originalValue)) { // Deffer filling in the field.
-            deferredValueConsumer.accept(originalValue);
+    public <C> Unprocessed clone(DeepCloningUtils deepCloningUtils, C original, C clone) {
+        Object originalValue = AbstractFieldCloner.getGenericFieldValue(original, field);
+        if (deepClone(deepCloningUtils, original.getClass(), originalValue)) { // Defer filling in the field.
+            return new Unprocessed(clone, field, originalValue);
         } else { // Shallow copy.
-            FieldCloner.setFieldValue(clone, field, originalValue);
+            AbstractFieldCloner.setGenericFieldValue(clone, field, originalValue);
+            return null;
         }
     }
 
-    private static boolean isDeepCloneField(DeepCloningUtils deepCloningUtils, Field field, Class<?> fieldInstanceClass,
-            Object originalValue) {
+    /**
+     * Obtaining the decision on whether or not to deep-clone is expensive.
+     * This method exists to cache those computations as much as possible,
+     * while maintaining thread-safety.
+     *
+     * @param deepCloningUtils never null
+     * @param fieldTypeClass never null
+     * @param originalValue never null
+     * @return true if the value needs to be deep-cloned
+     */
+    private boolean deepClone(DeepCloningUtils deepCloningUtils, Class<?> fieldTypeClass, Object originalValue) {
         if (originalValue == null) {
             return false;
         }
-        return deepCloningUtils.getDeepCloneDecision(field, fieldInstanceClass, originalValue.getClass());
+        /*
+         * This caching mechanism takes advantage of the fact that, for a particular field on a particular class,
+         * the types of values contained are unlikely to change and therefore it is safe to cache the calculation.
+         * In the unlikely event of a cache miss, we recompute.
+         */
+        boolean isValueDeepCloned = valueDeepCloneDecision.updateAndGet(old -> {
+            Class<?> originalClass = originalValue.getClass();
+            if (old == null || old.clz != originalClass) {
+                return new Metadata(originalClass, deepCloningUtils.isClassDeepCloned(originalClass));
+            } else {
+                return old;
+            }
+        }).decision;
+        if (isValueDeepCloned) { // The value has to be deep-cloned. Does not matter what the field says.
+            return true;
+        }
+        /*
+         * The decision to clone a field is constant once it has been made.
+         * The fieldTypeClass is guaranteed to not change for the particular field.
+         */
+        if (fieldDeepCloneDecision.get() < 0) {
```

Comment:
This could be replaced by `AtomicBoolean` with a null check.

Reply:
AtomicBoolean doesn't have a `null` value. And if I didn't make the field final, it wouldn't be thread-safe anymore.

Reply:
I missed the `final` field; the `AtomicInteger` is fine then.